signer/jws.go

```diff
@@ -0,0 +1,480 @@
+		delete(data, "alg")
+	}
+
+	return nil
+}
+
+func populateTime(data map[string]interface{}, s string, holder *time.Time) error {
+	if val, ok := data[s]; ok {
+		if value, err := time.Parse(time.RFC3339, val.(string)); err != nil {
+			return MalformedSignatureError{msg: fmt.Sprintf("failed to parse time for %s attribute, it's not in RFC3339 format", s)}
+		} else {
+			*holder = value
+			delete(data, s)
+		}
+	}
+	return nil
+}
+
+func populateExtendedAttributes(data map[string]interface{}, critical []string, holder *[]Attribute) error {
+	extendedAttr := make([]Attribute, 0)
+	for _, val := range critical {
+		extendedAttr = append(extendedAttr, Attribute{
+			Key:      val,
+			Critical: true,
+			Value:    data[val],
+		})
+		delete(data, val)
+	}
+	delete(data, headerKeyCrit)
+
+	for key, val := range data {
+		extendedAttr = append(extendedAttr, Attribute{
+			Key:      key,
+			Critical: false,
+			Value:    val,
+		})
+	}
+
+	*holder = extendedAttr
+	return nil
+}
+
+func getAndValidateCriticalHeaders(pHeaders map[string]interface{}) ([]string, error) {
+	// using map for performance and that's the reason all values are bool true
+	headersMarkedCrit := map[string]bool{headerKeySigningTime: true}
+	if _, ok := pHeaders[headerKeyExpiry]; ok {
+		headersMarkedCrit[headerKeyExpiry] = true
+	}
+
+	var crit []string
+	if val, ok := pHeaders[headerKeyCrit]; ok {
+		critical := reflect.ValueOf(val)
+		for i := 0; i < critical.Len(); i++ {
+			var val = critical.Index(i).Interface().(string)
+			if _, ok := headersMarkedCrit[val]; ok {
+				delete(headersMarkedCrit, val)
+			} else {
+				crit = append(crit, val)
+			}
+		}
+
+		if len(headersMarkedCrit) != 0 {
+			// This is not taken care by VerifySignerInfo method
+			return crit, MalformedSignatureError{"required headers not marked critical"}
+		}
+		return crit, nil
+	} else {
+		// This is not taken care by VerifySignerInfo method
+		return crit, MalformedSignatureError{"missing `crit` header"}
+	}
+}
+
+func getSignedAttrs(req SignRequest, method jwt.SigningMethod) map[string]interface{} {
+	attrs := make(map[string]interface{})
+	attrs[headerKeyAlg] = method.Alg()
+	attrs[headerKeySigningTime] = req.SigningTime.Format(time.RFC3339)
+	var crit = []string{headerKeySigningTime}
+	if !req.Expiry.IsZero() {
+		attrs[headerKeyExpiry] = req.Expiry.Format(time.RFC3339)
+		crit = append(crit, headerKeyExpiry)
+	}
+	attrs[headerKeyCty] = req.PayloadContentType
+
+	for _, elm := range req.ExtendedSignedAttrs {
+		attrs[elm.Key] = elm.Value
+		if elm.Critical {
+			crit = append(crit, elm.Key)
+		}
+	}
+
+	attrs[headerKeyCrit] = crit
+	return attrs
+}
+
+// ***********************************************************************
+// jwsEnvelope-JSON specific code
+// ***********************************************************************
+const (
+	// JWSPayloadContentType describes the media type of the jwsEnvelope envelope.
+	JWSPayloadContentType = "application/vnd.cncf.notary.v2.jws.v1"
```

Comment:
This shouldn't be exported, correct?

Reply:
Its intended to be exported as caller can use it for specifying the contentType during sign operation.